Feature add google ads plugins

maciejSamerdak opened the pull request.

kbeker commented on company_website/static/common/scripts/gtag_setup.js:
```diff
@@ -0,0 +1,6 @@
+window.dataLayer = window.dataLayer || [];
+function gtag(){
+  dataLayer.push(arguments);
+}
+gtag('js', new Date());
+gtag('config', 'AW-607460288');
```
Is this script anywhere called?

maciejSamerdak replied:
Right below. You were wondering why it accepts arguments if none are declared in the definition? Well how about this:
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/arguments

Turns out it isn't obligatory to declare function arguments in the JS. 😛

kbeker replied:
![Uploading what-kind-of-black-magic-sorcery-is-this.jpg…]()

kbeker replied:
![what-kind-of-black-magic-sorcery-is-this](https://user-images.githubusercontent.com/26469852/90101204-8b1f0c00-dd3e-11ea-8489-bf44404f9a49.jpg)

maciejSamerdak replied:
IKR? 😁

kbeker commented on company_website/static/estimate_project/scripts/conversion_tracking.js:
```diff
@@ -0,0 +1,16 @@
+        'event',
+        'conversion',
+        {
+            'send_to': 'AW-607460288/pqT6CMvlzdkBEMC31KEC',
```
I'm wondering if this should be a part of django settings because it is somehow production parameters

maciejSamerdak replied:
So far the cleanest way in my opinion I've found so far would be to declare a variable in the head template, after passing the constant to the context:
```
%script
    window.conversion_tracking_target_address = {{ conversion_tracking_target_address }}
```
Then I can access that from the .js files. Would this solution be acceptable?

kbeker commented on company_website/static/common/scripts/gtag_setup.js:
```diff
@@ -0,0 +1,6 @@
+  dataLayer.push(arguments);
+}
+gtag('js', new Date());
+gtag('config', 'AW-607460288');
```
Same here about `AW-607460288` parameter. It should be in my opinion be injected here instead of hardcoded here

maciejSamerdak commented:
I've changed the approach and moved the scripts to the head, like it's apparently advised by google and everybody else. I also moved those Google Ads IDs to the production settings' file. Also fixed few bugs I encountered.

kbeker commented on company_website/templates/common_partials/head.haml:
```diff
@@ -9,6 +9,14 @@
+/ Global site tag (gtag.js) - Google Ads: 607460288
+%script{:async => "", :src => "https://www.googletagmanager.com/gtag/js?id={{ google_ads_conversion_id }}"}
+:javascript
+  window.dataLayer = window.dataLayer || [];
+  function gtag(){ dataLayer.push(arguments); }
+  gtag('js', new Date());
+  gtag('config', '{{ google_ads_conversion_id }}');
```
I saw that you delete `gtag_setup.js` file and you said that it should be in header. It can be in header but it should never be just javascript code in haml/html file. Please move it to file again. I know that you can here init variable with `google_ads_conversion_id ` setting and use it in file and please try to do this.

maciejSamerdak replied:
Done.

kbeker commented on company_website/view_helpers.py:
```diff
@@ -2,11 +2,21 @@
-class CustomTemplateView(TemplateView):
+class GoogleAdsMixin(ContextMixin):
+    extra_context = {
+        "google_ads_conversion_id": GOOGLE_ADS_CONVERSION_ID,
+        "google_ads_conversion_target_address": GOOGLE_ADS_CONVERSION_TARGET_ADDRESS,
```
I think that we still also pass it as capital letters instead of small ones. Please change it. These variables are still our settings and they should be easy to recognize

maciejSamerdak replied:
Done

kbeker commented on company_website/templates/base.haml:
```diff
@@ -6,12 +6,12 @@
 {% block extra_head %}
 {% endblock %}
 
+.google-analytics
+    %img.google-analytics-image{:src => "{% google_analytics %}", :alt => "Google Analytics"}
+
```
Why you moved this?

maciejSamerdak replied:
**.google_analytics** element was causing extra head content to be placed in the body.
I can't get this to work otherwise: **haml** doesn't allow div to be placed in head, at least with current structure. Should we use **%head** tag explicitly, if at all possible? But even then I can't tell it will work, because when I place it in the **extra_head** block, it just vanishes altogether. 😕

maciejSamerdak replied:
I just checked and even the explicit use of **%head** and **%body** tags doesn't help: the div is forced into the body anyway!

It is ill-advised to put divs in the head anyway, as it should be reserved for a meta-like data only. It has been documented by W3C for example.

Looks like everything is working against putting those elements in the head. 😛

kbeker commented on company_website/templates/estimate_project.haml:
```diff
@@ -10,6 +10,12 @@
 {% block extra_head %}
     {% stylesheet 'estimate_project' %}
+    :javascript
+      function gtag_report_conversion(url) {
+        var callback = function () { if (typeof(url) != 'undefined') { window.location = url; } };
+        gtag('event', 'conversion', { 'send_to': '{{ google_ads_conversion_target_address }}', 'event_callback': callback });
+        return false;
+      }
```
It should be also in file

maciejSamerdak replied:
Done.

kbeker commented on company_website/templates/estimate_project.haml:
```diff
@@ -10,6 +10,12 @@
 {% block extra_head %}
     {% stylesheet 'estimate_project' %}
+    :javascript
+      function gtag_report_conversion(url) {
```
Please also remember that in JS is convention to name methods/variables etc. with camelCase instead of snake_case

maciejSamerdak replied:
Well apparently Google thinks otherwise, since this code comes entirely from their official instruction... 🤷

maciejSamerdak replied:
Done. I've changed it, because it appears to be called only locally, via the onclick action of the tracked elements.
Unless Google likes to practice some obscure, cryptic bullsh*t like something you'd expect from Apple...
... do they? 😨

kbeker replied:
![xd-696x456](https://user-images.githubusercontent.com/26469852/90410091-3e269700-e0aa-11ea-815c-8da26ea5cc3f.jpg)

kbeker commented on company_website/templates/estimate_project/estimate_project_form.haml:
```diff
@@ -63,7 +63,7 @@
               %a{:href => "{% url 'privacy_and_policy' %}"}  Privacy Policy
             {% url "thank_you" as thank_you_url %}
-            %input.custom-button{:type => "submit", :value => "Estimate Project", :class => "estimate-button", :id => "btnSubmit", :onclick => "return gtag_report_conversion('{{ thank_you_url }}')"}
+            %input.custom-button{:type => "submit", :value => "Estimate Project", :class => "estimate-button", :id => "btnSubmit", :onclick => 'return gtag_report_conversion("https://{{ request.get_host }}{{ thank_you_url }}")'}
```
`https` - It won't be always with ssl because locally and on test server we don't have any cert. We have such variable in our settings like `URL_PREFIX`. Please use this variable to differentiate correct prefix

maciejSamerdak replied:
It doesn't really matter anyway, because it will only work on the production, or so we're told.
Guess I could've just as well hardcoded **https://codepoets** 🤷

maciejSamerdak replied:
Done, I've passed the URL_PREFIX to the context data.

maciejSamerdak commented on company_website/view_helpers.py:
```diff
@@ -2,11 +2,21 @@
 
 from company_website.models import PageSeo
+from project_liberation.settings.production import GOOGLE_ADS_CONVERSION_ID
+from project_liberation.settings.production import GOOGLE_ADS_CONVERSION_TARGET_ADDRESS
 
 
-class CustomTemplateView(TemplateView):
+class GoogleAdsMixin(ContextMixin):
+    extra_context = {
+        "GOOGLE_ADS_CONVERSION_ID": GOOGLE_ADS_CONVERSION_ID,
+        "GOOGLE_ADS_CONVERSION_TARGET_ADDRESS": GOOGLE_ADS_CONVERSION_TARGET_ADDRESS,
```
Please tell me, is this acceptable? Because I have this nasty feeling it isn't.
You told me to move those values to the production settings, but these settings are not defined in the base settings file. Should I therefore:
1. Define those settings in the base or development settings and assign them some bogus values?
2. Move them to the base settings?
3. It is completely acceptable to access them here like this and I should shut up. 😛

kbeker replied:
Right, first, we should create this options also in base.py file but there they should be commented out. Second this is to create some mock ids/settings in development.py. It should work then :) Is it nasty solution? I think we haven't better right now and we should keep with this one :)

maciejSamerdak replied:
Done.

kbeker commented on company_website/static/common/scripts/head/google_tag_manager.js:
```diff
@@ -0,0 +1,10 @@
+(function(w, d, s, l, i){
```
Please at least use variables shortcuts instead of one letter naming

maciejSamerdak replied:
Crikey, I don't even no what half of them are! They just give you that nasty HTML-nested minimised script and I can't find any documentation or any kind of breakdown for it anywhere! 😩 
It's all this "Copy and paste rady, bro! No need to know what's all this for, trust me!" BS

kbeker replied:
They are probably from left to right: window, document, script, layer, id :)

maciejSamerdak replied:
Yeah, super silly of me not to check the function call at the bottom. 😖

kbeker commented on company_website/static/common/scripts/head/google_tag_manager.js:
```diff
@@ -0,0 +1,10 @@
+    w[l].push({'gtm.start': new Date().getTime(), event: 'gtm.js'});
+    var f = d.getElementsByTagName(s)[0],
+    j = d.createElement(s), dl = l != 'dataLayer' ? '&l=' + l: '';
+    j.async=true;
+    j.src='https://www.googletagmanager.com/gtm.js?id=' + i + dl;
```
Missing spaces between equal sign

maciejSamerdak replied:
Done.
